src/senaite/core/upgrade/v02_03_000.py

```diff
@@ -411,3 +421,71 @@ def fixed_point_value_to_string(value, precision):
     str_value = template % (sign, front, fra)
     # strip off trailing zeros and possible dot
     return str_value.rstrip("0").rstrip(".")
+
+
+def add_isanalyte_index(portal):
+    cat = api.get_tool(ANALYSIS_CATALOG)
+    logger.info("Add isAnalyte index to {} ...".format(cat.id))
+    if add_catalog_index(cat, "isAnalyte", "", "BooleanIndex"):
+        reindex_catalog_index(cat, "isAnalyte")
+    logger.info("Add isAnalyte index to {} [DONE]".format(cat.id))
+
+
+def mark_retracted_analyses(portal):
+    """Sets the IRetracted interface to analyses that were retracted
+    """
+    logger.info("Applying IRetracted interface to retracted analyses ...")
+    query = {
+        "portal_type": ["Analysis", "ReferenceAnalysis", "DuplicateAnalysis"]
```

Comment:
Please search here for `"review_state": "retracted"`, or for both `"review_state": ["retracted", "rejected"]` and merge with the function below.

Reply:
I got now why you did the upgrade step like this, because the current state can be different due to re-activation and you would like to mark any analyses that have ever been rejected / retracted.
However, I think we should do a `noLonverProvides` if this happened, because it does not make sense to keep it, because these are then no longer rejected/retracted.

Reply:
Yes, I did not filter by review state because I don't know if there are analyses in a further status passed retracted/rejected. I thought the same, that shouldn't make sense. 

However, I later I realized that client-specific add-ons (not my case) may exist that do consider additional states past reject/retract, specially after the "dispacth" transition and/or senaite.storage were added. And I left the query w/o changes.

Still, filtering by review-state won't have any impact on any of the instances using the add-ons I am aware of.

Reply:
Yes, that dispatch state...

Not sure if it was a good idea, but it might make sense when using senaite.referral or when subcontracting external labs for analysis.

Nevertheless, I would change the upgrade step to either iterate once over all analyses or consider only final rejected/retracted states.

The reason is that we have here instances with 500k Analyses and it is a pain to iterate through all of them for migration (takes 2.5 days currently)

Reply:
Sure, let's keep things simpler and faster https://github.com/senaite/senaite.core/pull/2120/commits/81aa786454859b8a0795c4b69d2a58a23ee9f487